manager/sql/databases.go

```diff
@@ -0,0 +1,276 @@
+							template varchar(1023) NOT NULL,
+							policy   varchar(255) NOT NULL,
+							FOREIGN KEY (policy) REFERENCES ladon_policy(id) ON DELETE CASCADE
+						)`,
+					},
+					Down: []string{
+						"DROP TABLE ladon_policy",
+						"DROP TABLE ladon_policy_subject",
+						"DROP TABLE ladon_policy_permission",
+						"DROP TABLE ladon_policy_resource",
+					},
+				},
+				{
+					Id: "2",
+					Up: []string{
+						`CREATE TABLE IF NOT EXISTS ladon_subject (
+							id          varchar(64) NOT NULL PRIMARY KEY,
+							has_regex   bool NOT NULL,
+							compiled    varchar(511) NOT NULL UNIQUE,
+							template    varchar(511) NOT NULL UNIQUE
+						)`,
+						`CREATE TABLE IF NOT EXISTS ladon_action (
+							id          varchar(64) NOT NULL PRIMARY KEY,
+							has_regex   bool NOT NULL,
+							compiled    varchar(511) NOT NULL UNIQUE,
+							template    varchar(511) NOT NULL UNIQUE
+						)`,
+						`CREATE TABLE IF NOT EXISTS ladon_resource (
+							id          varchar(64) NOT NULL PRIMARY KEY,
+							has_regex   bool NOT NULL,
+							compiled    varchar(511) NOT NULL UNIQUE,
+							template    varchar(511) NOT NULL UNIQUE
+						)`,
+						`CREATE TABLE IF NOT EXISTS ladon_policy_subject_rel (
+							policy   varchar(255) NOT NULL,
+							subject  varchar(64) NOT NULL,
+							PRIMARY KEY (policy, subject),
+							FOREIGN KEY (policy) REFERENCES ladon_policy(id) ON DELETE CASCADE,
+							FOREIGN KEY (subject) REFERENCES ladon_subject(id) ON DELETE CASCADE
+						)`,
+						`CREATE TABLE IF NOT EXISTS ladon_policy_action_rel (
+							policy  varchar(255) NOT NULL,
+							action  varchar(64) NOT NULL,
+							PRIMARY KEY (policy, action),
+							FOREIGN KEY (policy) REFERENCES ladon_policy(id) ON DELETE CASCADE,
+							FOREIGN KEY (action) REFERENCES ladon_action(id) ON DELETE CASCADE
+						)`,
+						`CREATE TABLE IF NOT EXISTS ladon_policy_resource_rel (
+							policy    varchar(255) NOT NULL,
+							resource  varchar(64) NOT NULL,
+							PRIMARY KEY (policy, resource),
+							FOREIGN KEY (policy) REFERENCES ladon_policy(id) ON DELETE CASCADE,
+							FOREIGN KEY (resource) REFERENCES ladon_resource(id) ON DELETE CASCADE
+						)`,
+					},
+					Down: []string{},
+				},
+				{
+					Id: "3",
+					Up: []string{
+						"CREATE INDEX ladon_subject_compiled_idx ON ladon_subject (compiled text_pattern_ops)",
+						"CREATE INDEX ladon_permission_compiled_idx ON ladon_action (compiled text_pattern_ops)",
+						"CREATE INDEX ladon_resource_compiled_idx ON ladon_resource (compiled text_pattern_ops)",
+					},
+					Down: []string{
+						"DROP INDEX ladon_subject_compiled_idx",
+						"DROP INDEX ladon_permission_compiled_idx",
+						"DROP INDEX ladon_resource_compiled_idx",
+					},
+				},
+			},
+		},
+		QueryInsertPolicy:             `INSERT INTO ladon_policy(id, description, effect, conditions) SELECT $1::varchar, $2, $3, $4 WHERE NOT EXISTS (SELECT 1 FROM ladon_policy WHERE id = $1)`,
```

Comment:
Do you know why this (`::varchar`) was previously not necessary?

Reply:
The PostgreSQL driver's type inference was returning two different types for the first and second instances of `$1` in the prepared statement, so the `::varchar` settled that. It was originally unnecessary because `$1` wasn't reused later in this query, the ID value was passed in twice.

Reply:
Thank you!